p2p-tests/lib/p2p.ts

```diff
@@ -0,0 +1,304 @@
+import * as fs from "fs-extra";
+import * as path from "path";
+import * as stream from "stream";
+import * as util from "util";
+import execa from "execa";
+import { StringDecoder } from "string_decoder";
+
+const PADDING = 12;
+const RESET = "\x1b[0m";
+
+const BLUE = "\x1b[34m";
+const CYAN = "\x1b[36m";
+const GREEN = "\x1b[32m";
+const MAGENTA = "\x1b[35m";
+const RED = "\x1b[31m";
+
+const colors = [BLUE, CYAN, GREEN, MAGENTA, RED];
+const assignedColors = {};
+
+const ROOT_PATH = path.join(__dirname, "..", "..");
+const P2P_TEST_PATH = path.join(ROOT_PATH, "p2p-tests");
+const BIN_PATH = path.join(ROOT_PATH, "target", "debug");
+
+const processes = [];
+
+export function cleanUpProcesses() {
+  processes.forEach(node => node.kill("SIGKILL"));
+}
+
+export function workspacePath(paths?: string[]) {
+  return path.join(ROOT_PATH, "p2p-tests", "workspace", ...paths);
+}
+
+export function keyPath(filename: string) {
+  return path.join(P2P_TEST_PATH, "keys", filename);
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function assertEqual(message: string, got: any, expected: any) {
+  log(`ASSERT: ${message}`);
+  strictAssert.deepStrictEqual(got, expected);
+  log("got what was expected:");
+  prettyPrint(got);
+}
+
+function prefix(pfx: string) {
+  if (assignedColors[pfx] === undefined) {
+    assignedColors[pfx] = colors.pop();
+  }
+
+  return `${RESET + assignedColors[pfx] + pfx.padEnd(PADDING) + RESET} | `;
+}
+
+export function log(message?: string) {
+  if (message) {
+    console.log(`${prefix("host")}${message}`);
+  } else {
+    console.log();
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function prettyPrint(obj: any) {
+  console.log(
+    util
+      .inspect(obj, { depth: null, colors: true })
+      .split("\n")
+      .map(line => " ".repeat(PADDING + 3) + line)
+      .join("\n")
+  );
+}
+
+// A transform that prefixes each line from the source with the given
+// string and pushes it to the sink.
+export class LinePrefix extends stream.Transform {
+  private buffer: string = "";
+  private stringDecoder = new StringDecoder();
+  constructor(private prefix: string) {
+    super();
+  }
+
+  _transform(data: Buffer, _encoding: string, next: () => void) {
+    const str = this.buffer + this.stringDecoder.write(data);
+    const lines = str.split(/\r?\n/);
+    this.buffer = lines.pop() || "";
+    lines.forEach(line => {
+      if (line === "") {
+        this.push("\n");
+      } else {
+        this.push(`${this.prefix}${line}\n`);
+      }
+    });
+    next();
+  }
+
+  _flush(done: () => void) {
+    this.push(`${this.prefix}${this.buffer}${this.stringDecoder.end()}\n`);
+    done();
+  }
+}
+
+export function initPeer(namespace: string) {
+  log(`init ${namespace}`);
+  log();
+
+  const radHome = workspacePath([namespace]);
+  fs.removeSync(radHome);
+  fs.mkdirsSync(radHome);
+
+  const result = runInNamespace(namespace, [
```

Comment:
No need to run this in a namespace. The command doesn’t do any networking.

Reply:
f2bd600831576367ff9ad16306881a1c4a8dc7e0